source/opt/ir_builder.h

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_OPT_IR_BUILDER_H_
+#define LIBSPIRV_OPT_IR_BUILDER_H_
+
+#include "opt/basic_block.h"
+#include "opt/instruction.h"
+#include "opt/ir_context.h"
+
+namespace spvtools {
+namespace opt {
+
+// In SPIR-V, ids are encoded as uint16_t, this id is guarantied to be always
+// invalid.
+constexpr uint32_t kInvalidId = (uint32_t)-1;
+
+// Helper class to abstract instruction construction and insertion.
+// |AnalysisToPreserve| asks the InstructionBuilder to preserve requested
+// analysis.
+// Supported analysis:
+//   - Def-use analysis
+template <ir::IRContext::Analysis AnalysisToPreserve =
+              ir::IRContext::kAnalysisNone>
+class InstructionBuilder {
+  static_assert(!(AnalysisToPreserve & ~ir::IRContext::kAnalysisDefUse),
+                "Only Def-use analysis update is supported");
+
+ public:
+  using InsertionPointTy = spvtools::ir::BasicBlock::iterator;
+
+  InstructionBuilder(ir::IRContext* context, InsertionPointTy insert_before)
+      : context_(context), insert_before_(insert_before) {}
+
+  // Creates a new selection merge instruction.
+  // The id |merge_id| is the merge basic block id.
+  ir::Instruction* AddSelectionMerge(uint32_t merge_id) {
+    std::unique_ptr<ir::Instruction> new_branch_merge(new ir::Instruction(
+        GetContext(), SpvOpSelectionMerge, 0, 0,
+        {{spv_operand_type_t::SPV_OPERAND_TYPE_ID, {merge_id}},
+         {spv_operand_type_t::SPV_OPERAND_TYPE_LITERAL_INTEGER, {0}}}));
```

Comment:
This should be:
{SPV_OPERAND_TYPE_SELECTION_CONTROL, {SpvSelectionControlMaskNone}}

I would also suggest changing the control to an argument and defaulting it to none.

Reply:
I agree, this should be an argument